### PR TITLE
UFAL/S3-CESNET direct downloads

### LINKED ISSUE #884: UFAL/S3 direct downloads
The original issue: https://github.com/ufal/clarin-dspace/issues/1153

- Get know how Cesnet allows the S3 direct downloads
    - pre-signed URLs
    - GROK: `In summary, CESNET allows S3 direct downloads by supporting pre-signed URLs, leveraging its S3-compatible protocol with a custom endpoint and bucket naming convention.`
- Think about tests, how we will test it
- Consider downloading the big files using nginx

CESNET direct downloads:
This command `aws s3 presign s3://testbucket/eighty-eight/10/27/80/102780081019350462616804225709692058487 --expires-in 3600 --endpoint-url <EDNPOINT> --region <REGION>`

Generates pre-shared url: `https://s3.cl4.du.cesnet.cz/testbucket/eighty-eight/10/27/80/102780081019350462616804225709692058487?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AG9WI0FABHBCRYV22F3N%2F20250226%2Feu-central-1%2Fs3%2Faws4_request&X-Amz-Date=20250226T140301Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&X-Amz-Signature=defccca5ba5a416a27f50a05c63329c147a30660172132698fac02226fb2d78`

TODO:
1. Think about tests
- We must call directly the CESNET API
- The best solution will be some python Integration tests which will upload some files and test edge case scenarios

BE: 
- return an Object with:
   - URL
   - Bitstream name (must contain type e.g., *png, *pdf,...)

### PR DESCRIPTION
**Pull request by milanmajchrak**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  0  |  0  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |
## Summary by CodeRabbit

- **New Features**
  - Added support for direct downloads of bitstreams from S3 storage using presigned URLs, enabling faster and more secure file access.
  - Introduced configuration options to enable S3 direct download and set URL expiration time.
- **Bug Fixes**
  - Improved error handling for missing S3 objects and misconfiguration during direct download attempts.
- **Tests**
  - Added unit and integration tests to verify correct presigned URL generation, redirection behavior, and error scenarios.

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java [hunk 1]
```diff
@@ -1335,4 +1342,52 @@ public void testChecksumLinkRepository() throws Exception {
...
+                    .andExpect(header().string("Location", containsString("https://")))
+                    .andExpect(header().string("Location", containsString("response-content-disposition")))
+                    .andExpect(header().string("Location", containsString(
+                            URLEncoder.encode("filename=\"" + fileName + "\"", StandardCharsets.UTF_8))));
+        } finally {
+            // Clean up the S3BitStoreService
+            s3BitStoreService = null;
+            configurationService.setProperty("s3.download.direct.enabled", false);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (milanmajchrak) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/BitstreamRestControllerIT.java]: done

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3DirectDownloadServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        this.s3Client = s3BitStoreService.s3Service;
+        try {
+            if (this.s3Client == null) {
+                this.s3BitStoreService.init();
+                this.s3Client = s3BitStoreService.s3Service;
+                if (s3Client == null) {
+                    throw new RuntimeException("S3 client wasn't initialized in the S3BitStoreService.");
+                }
+            }
+        } catch (Exception e) {
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3DirectDownloadServiceImpl.java]: the catch can be in from of s3 client, you will avoid cathing runtimexception

**Review by vidiecan — CHANGES_REQUESTED** [review]
